src/Workspaces/Core/Portable/EmbeddedLanguages/Common/EmbeddedSyntaxToken.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.Common
+{
+    internal struct EmbeddedSyntaxToken<TSyntaxKind> where TSyntaxKind : struct
+    {
+        public readonly TSyntaxKind Kind;
+        public readonly ImmutableArray<EmbeddedSyntaxTrivia<TSyntaxKind>> LeadingTrivia;
+        public readonly ImmutableArray<VirtualChar> VirtualChars;
+        public readonly ImmutableArray<EmbeddedSyntaxTrivia<TSyntaxKind>> TrailingTrivia;
+        internal readonly ImmutableArray<EmbeddedDiagnostic> Diagnostics;
+        public readonly object Value;
+
+        public EmbeddedSyntaxToken(
+            TSyntaxKind kind,
+            ImmutableArray<EmbeddedSyntaxTrivia<TSyntaxKind>> leadingTrivia,
+            ImmutableArray<VirtualChar> virtualChars,
+            ImmutableArray<EmbeddedSyntaxTrivia<TSyntaxKind>> trailingTrivia,
+            ImmutableArray<EmbeddedDiagnostic> diagnostics, object value)
+        {
+            Kind = kind;
+            LeadingTrivia = leadingTrivia;
+            VirtualChars = virtualChars;
+            TrailingTrivia = trailingTrivia;
+            Diagnostics = diagnostics;
+            Value = value;
+        }
+
+        public bool IsMissing => VirtualChars.IsEmpty;
```

Comment:
Is it safe to call `.IsEmpty` on a default `ImmutableArray`?

Reply:
In a SyntaxToken none of these will be empty. I will add checks for that in the constructor.

Reply:
#resolved.